stan-dev/cmdstan
# Fixing build issues

#### Submisison Checklist


#### Summary:

Fixes a lot of build issues.

This allows users in Windows, Linux, and Mac to run CmdStan without additional configuration. The defaults allow models to be built without additional configuration.

#### Intended Effect:

Fixes build issues that users were reporting:
- #586 
- #590 
- #588 

#### How to Verify:

Build from:
- Mac, Windows, and Linux
- from a fresh clone without a `make/local` or `~/.config/cmdstan/make.local` file present

And build the bernoulli example.

Please verify that this works on your local machine or whatever other machine you have access to.

#### Side Effects:

These are the known side-effects in this pull request. These shouldn't affect users. The second one may affect developers:
- Mac: on develop, we have included in the `CXXFLAGS` `-stdlib=libc++`. In this branch, this is omitted. 
    Without it, it builds fine on my machine. I removed the default compiler from being set to `clang++`. Instead, we use the system default, which usually defaults to `c++`. The math library's `detect_cc` searches for the string `clang++` and uses that to set options.
- Linux: tests don't build with the default values. In order to build them, you need to include `GTEST_CXXFLAGS += -DGTEST_HAS_PTHREAD=0` in `make/local` or `~/.config/cmdstan/make.local`.
    This is also happening because `CXX`'s default value is `c++`. The `os_linux` file uses the value there (indirectly) to add to the `GTEST_CXXFLAGS`. We should probably just add that directly for all of Linux, but the fix should be in the Math library. This has been documented.

#### Documentation:

- Added subsection mentioning how to build tests in linux. 
- Minor updates to change `CC` to `CXX` (this may conflict with the other PR, but we can work through that easily).

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Generable

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
